### ISSUE TITLE
Custom serialization of body objects

### ISSUE THREAD

**Issue by MarcelBochtler** [SMILEY4/ktor-openapi-tools#3]
I have an endpoint which expects this body:
```json
{
	"name": "entity name",
	"description": "entity description"
}
```

In the application, this is deserialized into
```kt
@Serializable
data class UpdateOrganization(
    val name: OptionalValue<String> = OptionalValue.Absent,
    val description: OptionalValue<String?> = OptionalValue.Absent
)
```

By using this custom serializer for `kotlinx.serialization`:
```kt
class OptionalValueSerializer<T>(private val valueSerializer: KSerializer<T>) : KSerializer<OptionalValue<T>> {
    override val descriptor: SerialDescriptor = valueSerializer.descriptor

    override fun deserialize(decoder: Decoder): OptionalValue<T> {
        val value = valueSerializer.deserialize(decoder)
        return OptionalValue.Present(value)
    }

    override fun serialize(encoder: Encoder, value: OptionalValue<T>) {
        when (value) {
            is OptionalValue.Absent -> {}
            is OptionalValue.Present -> valueSerializer.serialize(encoder, value.value)
        }
    }
}
```

The generated spec is displayed like this in the Swagger UI:
```json
{
  "name": {
    "value": "entity name"
  },
  "description": {
    "value": "entity description"
  }
}

```

Is there a way to change the serialization for this object when generating the OpenApi spec?

**Comment by SMILEY4** [comment]
Hi,

the library i'm using to generate the json-schemas is using jackson (and does not support kotlinx-serialization afaik).

As a workaround you could provide the json-schemas for the body yourself. See the example `ExternalJsonSchemaExample.kt` on how to do that.
With this, you could use a different library that supports kotlinx-serializsation (quick google search returned this: https://github.com/Ricky12Awesome/json-schema-serialization).
With this, you could then specify your body like this:
```kotlin
request {
	body("/schemas/UpdateOrganization")
}
```
... and provide a custom endpoint returning the json-schema:
```kotlin
route("schemas") {
    get("UpdateOrganization") {
        call.respond(HttpStatusCode.OK, globalJson.encodeToSchema(UpdateOrganization.serializer(), generateDefinitions = false))
    }
}
```
For me (with a hacky ""copy"" of your OptionalValue), it looks like this in swagger, however, its now not part of the single openapi-file anymore.
```json
{
  "name": "string",
  "description": "string"
}
```

Properly supporting kotlinx-serialization as an alternative is on my todo-list. I think it would also be a good idea to be able to provide your own "object-to-jsonSchema"-converter for more customisation.

**Comment by MarcelBochtler** [comment]
> As a workaround you could provide the json-schemas for the body yourself.

Yes this works.

However, due to this:

> its now not part of the single openapi-file anymore.

The generated spec won't add these external schemas to the `components` when using the option `schemasInComponentSection = true`.
This also means, that any generated clients won't contain these models.

**Comment by SMILEY4** [comment]
I just released a new version (0.6.0) with reworked/improved support for custom schemas. See https://github.com/SMILEY4/ktor-swagger-ui/wiki/Documenting-Routes-(Bodies)#custom-schemas for more information.

For your case, you could now use e.g. "Ricky12Awesome/json-schema-serialization" as a custom json-schema builder:

```kotlin
install(SwaggerUI) {
    schemas {
        jsonSchemaBuilder { type ->
            // build the json-schema for the given type
            globalJson.encodeToSchema(serializer(type), generateDefinitions = false)
        }
    }
}
// ... define bodies as usual
body<UpdateOrganization> { /*...*/ }
```
